File: src/graph/bellman-ford/km94.py
from typing import List, Tuple


vertexNum, edgeNum = map(int, input().strip().split(" "))

edges: List[Tuple[int, int, int]] = []

for _ in range(edgeNum):
    start, end, price = map(int, input().strip().split(" "))
    edges.append((start, end, price))

Infinity = (1 << 31) - 1
minDist = [Infinity for _ in range(vertexNum + 1)]

start = 1
end = vertexNum
minDist[start] = 0

for _ in range(vertexNum - 1):
    updated = False
    for edge in edges:
        start_, end_, price_ = edge

        if minDist[start_] != Infinity and minDist[end_] > minDist[start_] + price_:
            minDist[end_] = minDist[start_] + price_
            updated = True

    if not updated:
        break

if minDist[end] == Infinity:
    print("unconnected")
else:
    print(minDist[end])

File: src/graph/bellman-ford/km94.cc
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

int main() {
  int vertexNum, edgeNum;
  cin >> vertexNum >> edgeNum;

  auto edges = vector<vector<int>>();

  for (int i = 0; i < edgeNum; i++) {
    int start, end, price;
    cin >> start >> end >> price;
    auto item = vector<int>{start, end, price};
    edges.emplace_back(item);
  }

  auto minDist = vector<int>(vertexNum + 1, INT_MAX);
  const auto start = 1;
  const auto end = vertexNum;

  minDist[start] = 0;
  for (auto i = 0; i < vertexNum - 1; i++) {
    auto updated = false;
    for (auto &edge : edges) {
      auto from = edge[0];
      auto to = edge[1];
      auto price = edge[2];

      if (minDist[from] != INT_MAX && minDist[to] > minDist[from] + price) {
        updated = true;
        minDist[to] = minDist[from] + price;
      }
    }

    if (!updated) {
      break;
    }
  }

  if (minDist[end] == INT_MAX) {
    cout << "unconnected" << endl;
  } else {
    cout << minDist[end] << endl;
  }
}
